### PR TITLE
Revision of spidy internals

### PR DESCRIPTION
**Pull request by influx6**
A rewrite for spidy internal behaviour.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -48,143 +49,132 @@ Usage:
...
+	path, err := url.Parse(*link)
 	if err != nil {
+		fmt.Printf("Error: Unparsable URI: %s\n", *link)
+		os.Exit(-1)
 		return
 	}
 
+	// deadLinks provides a lists to contain all failed url links that are
+	// considered deadlinks.
+	var deadLinks []*linkReport
```
Comment (egonzo) [inline main.go:67]: When using a slice, there is no need to use the address of use the value instead. A slice is reference type that has a pointer to the underlying array.

Its not a big deal but its more idiomatic.

File: main.go [hunk 2]
```diff
@@ -48,143 +49,132 @@ Usage:
...
+			deadLinks = append(deadLinks, link)
 		}
 	}
 
 }
 
-// queueLinks is used for making http calls to the queued links in queue channel
+// collectFrom uses a recursive function to map out the needed lists of links to.
+// It returns a channel through which the acceptable links can be crawled from.
+func collectFrom(path *url.URL, doExternals bool) <-chan *linkReport {
```
Comment (egonzo) [inline main.go]: You can make the returning channel a value of linkReport.  

Such as: <-chan linkReport. The link report type only contains primitive values such as string and int.

File: main.go [hunk 3]
```diff
@@ -48,143 +49,132 @@ Usage:
...
+	res, err := httpClient.Get(path)
 	if err != nil {
-		fmt.Printf(`Host: %s
-URI: %s
-Error: %s
-
-`, host, uri, err.Error())
-
-		deadLinks = append(deadLinks, uri)
+		dead <- &linkReport{Link: path, Status: 404}
```
Comment (egonzo) [inline main.go:123]: You should send the actual status code not a hard coded 404.

**Comment by egonzo** [comment]
Have you run this new code yet?

**Comment by influx6** [comment]
Yes I ran the code and it worked, when all links where possible it ended well, the function collects, collects all links for the page and their respective siblinks, until it has exhausted the list then closes the channel

File: main.go [hunk 4]
```diff
@@ -48,143 +49,132 @@ Usage:
...
+		fmt.Printf("Error: Unparsable URI: %s\n", *link)
+		os.Exit(-1)
 		return
 	}
 
+	// deadLinks provides a lists to contain all failed url links that are
+	// considered deadlinks.
+	var deadLinks []*linkReport
+
+	dead := collectFrom(path, !*hostOnly)
```
Comment (influx6) [inline main.go:69]: The function collectFrom takes the path and pulls out the links for that page, using that first set it recursively crawls their sublinks until the master link list is done and it closes the channel, even if all were possible it is guaranteed to end once done

**Pull request merged**